pkg/loop/internal/chain_reader.go

```diff
@@ -0,0 +1,293 @@
+package internal
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"unicode"
+
+	"github.com/fxamacker/cbor/v2"
+	libocr "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+	"google.golang.org/grpc/status"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/loop/internal/pb"
+	"github.com/smartcontractkit/chainlink-relay/pkg/types"
+)
+
+var _ types.ChainReader = (*chainReaderClient)(nil)
+
+type chainReaderClient struct {
+	*brokerExt
+	grpc pb.ChainReaderClient
+}
+
+// enum of all known encoding formats for versioned data
+const (
+	JSONEncodingVersion = iota
+	CBOREncodingVersion
+)
+
+// Version to be used for encoding ( version used for decoding is determined by data received )
+// These are separate constants in case we want to upgrade their data formats independently
+const ParamsCurrentEncodingVersion = CBOREncodingVersion
+const RetvalCurrentEncodingVersion = CBOREncodingVersion
+
+func encodeVersionedBytes(data any, version int32) (*pb.VersionedBytes, error) {
+	var bytes []byte
+	var err error
+
+	switch version {
+	case JSONEncodingVersion:
+		bytes, err = json.Marshal(data)
+	case CBOREncodingVersion:
+		enco := cbor.CoreDetEncOptions()
+		enco.Time = cbor.TimeRFC3339Nano
+		var enc cbor.EncMode
+		enc, err = enco.EncMode()
+		if err != nil {
+			return nil, err
+		}
+		bytes, err = enc.Marshal(data)
+	default:
+		return nil, fmt.Errorf("unsupported encoding version %d for data %v", version, data)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", types.InvalidTypeError{}, err)
+	}
+	return &pb.VersionedBytes{Version: uint32(version), Data: bytes}, nil
+}
+
+func decodeVersionedBytes(res any, vData *pb.VersionedBytes) error {
+	var err error
+	switch vData.Version {
+	case JSONEncodingVersion:
+		err = json.Unmarshal(vData.Data, res)
+	case CBOREncodingVersion:
+		err = cbor.Unmarshal(vData.Data, res)
+	default:
+		return fmt.Errorf("unsupported encoding version %d for versionedData %v", vData.Version, vData.Data)
+	}
+
+	if err != nil {
+		return fmt.Errorf("%w: %w", types.InvalidTypeError{}, err)
+	}
+	return nil
+}
+
+func isArray(vData *pb.VersionedBytes) (bool, error) {
+	data := vData.Data
+	if len(data) > 0 {
+		switch vData.Version {
+		case JSONEncodingVersion:
+			i := int(0)
+			for ; i < len(data); i++ {
+				if !unicode.IsSpace(rune(data[i])) {
+					break
+				}
+			}
+			return i < len(data) && data[i] == '[', nil
```

Comment:
Can we use https://pkg.go.dev/strings#TrimLeftFunc ?

Reply:
The only problem with `TrimLeftFunc` is the same as with `TrimSpace` and `TrimPrefix`:  all of them accept only a `string` type as input but we have a raw `[]byte`.

We could convert the whole byte array into a string, and then use `TrimSpace` or one of the others on it.  But then we'd have to worry about character encodings and portability.  I guess by casting from `byte` to `rune` we're technically already supposed to worry about that, so maybe it's not really a big deal?  Another reason I went for this way was that we don't know how large this buffer might be.  Converting the entire buffer into a string seems like a waste if all we need to check is the first non-whitespace character.

Reply:
There is a bytes version https://pkg.go.dev/bytes#TrimLeftFunc